core/consensus/babe/impl/babe_synchronizer_impl.hpp

```diff
@@ -3,62 +3,100 @@
+        std::shared_ptr<blockchain::BlockTree> block_tree,
+        std::shared_ptr<BlockExecutor> block_executor,
+        std::shared_ptr<network::Router> router,
+        std::shared_ptr<libp2p::basic::Scheduler> scheduler,
+        std::shared_ptr<crypto::Hasher> hasher);
+
+    void enqueue(const primitives::BlockInfo &block_info,
+                 const libp2p::peer::PeerId &peer_id,
+                 SyncResultHandler &&handler) override;
 
-    void request(const primitives::BlockId &from,
-                 const primitives::BlockHash &to,
+    void enqueue(const primitives::BlockHeader &header,
                  const libp2p::peer::PeerId &peer_id,
-                 const BlocksHandler &block_list_handler) override;
+                 SyncResultHandler &&handler) override;
 
    private:
-    /**
-     * Select next client to be polled
-     * @param polled_clients clients that we already polled
-     * @return next clint to be polled
-     */
-    std::shared_ptr<network::SyncProtocolClient> selectNextClient(
-        std::unordered_set<std::shared_ptr<network::SyncProtocolClient>>
-            &polled_clients) const;
-    /**
-     * Request blocks from provided peers
-     * @param response_res block request message
-     * @param polled_clients peers that were already requested
-     * @param requested_blocks_handler handler of received blocks
-     */
-    void pollClients(network::BlocksRequest response_res,
-                     const libp2p::peer::PeerId &peer_id,
-                     const BlocksHandler &requested_blocks_handler) const;
-
-    std::shared_ptr<network::SyncClientsSet> sync_clients_;
-    log::Logger logger_;
-    const application::AppConfiguration &app_configuration_;
+    bool isInQueue(const primitives::BlockHash &hash) const;
+
+    void findCommonBlock(const libp2p::peer::PeerId &peer_id,
+                         primitives::BlockNumber lower,
+                         primitives::BlockNumber upper,
+                         primitives::BlockNumber hint,
+                         SyncResultHandler &&handler) const;
+
+    void loadBlocks(const libp2p::peer::PeerId &peer_id,
+                    primitives::BlockInfo from,
+                    SyncResultHandler &&handler);
+
+    void askNextPortionOfBlocks();
+    void applyNextBlock();
+    size_t discardBlock(const primitives::BlockHash &block);
+    void prune(const primitives::BlockInfo &finalized_block);
+
+    std::shared_ptr<blockchain::BlockTree> block_tree_;
+    std::shared_ptr<BlockExecutor> block_executor_;
     std::shared_ptr<network::Router> router_;
+    std::shared_ptr<libp2p::basic::Scheduler> scheduler_;
+    std::shared_ptr<crypto::Hasher> hasher_;
+
+    log::Logger log_ =
+        log::createLogger("BabeSynchronizer", "babe_synchronizer");
+
+    bool stops_ = false;
+
+    struct KnownBlock {
+      primitives::BlockData data;
+      std::set<libp2p::peer::PeerId> peers;
```

Comment:
Not quite sure what peers mean here. Is it the list of peers which are guarenteed to know this block?

Reply:
Yep. Comment will be added